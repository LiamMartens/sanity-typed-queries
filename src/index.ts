export { defineDocument, defineObject, defineFields } from './extractor'
export { QueryBuilder } from './query/builder'

export * from './schema/fields'
export * from './schema/validation'
export * from './types'
