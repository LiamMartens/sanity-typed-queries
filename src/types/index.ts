export type { Block } from './block'
export type { Image } from './image'

export const to = Symbol('type of object referred to by the reference')

interface FileAsset {
  url: string
  path: string
  size: number
  assetId: string
  mimeType: string
  sha1hash: string
  extension: string
  uploadId?: string
  originalFilename?: string
}

export interface File {
  _type: 'file'
  asset: Reference<FileAsset>
  [key: string]: any
}

export interface Geopoint {
  _type: 'geopoint'
  lat: number
  lng: number
  alt: number
}

export interface Slug {
  _type: 'slug'
  current: string
}

export interface Reference<Type = any> {
  _type: 'reference'
  _ref: string
  _weak?: boolean
  [to]: Type
}
