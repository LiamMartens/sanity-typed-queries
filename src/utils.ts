export function capitalise(text: string) {
  return text[0].toUpperCase() + text.slice(1)
}

export function splitStringByCase(text: string) {
  return capitalise(
    text
      .replace(/[A-Z]+[a-z]/g, h =>
        h.length > 2
          ? h.slice(0, -2) + ' ' + capitalise(h.slice(-2).toLowerCase())
          : Array(h).join(' ')
      )
      .replace(/[a-z][A-Z]+/g, h => `${h[0]} ${h.slice(1)}`)
  )
}

export function inArray<A>(items: A | A[]) {
  return Array.isArray(items) ? items : [items]
}

export const createProxy: (path: string[]) => any = (path: string[]) =>
  new Proxy(path, {
    get(target, prop: string) {
      if (prop === 'use') return () => target.join('.').replace('.->', '->')
      if (prop === 'resolve') {
        return (attr: string) => createProxy([...target, `->${attr}`])
      }
      return createProxy([...target, prop])
    },
  })
