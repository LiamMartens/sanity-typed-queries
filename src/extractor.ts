import { DocumentField, UnnamedField, FieldType } from './schema/fields'

import { splitStringByCase } from './utils'
import { RuleForRequired } from './schema/validation'
import { QueryBuilder } from './query/builder'

interface SanityDocumentCreator<T = never> {
  [key: string]: UnnamedField<T>
}

type RequiredField<T> = T
type OptionalField<T> = T | undefined

export type SanityDocumentObject<T extends SanityDocumentCreator<any>> = {
  [P in keyof T]: 'validation' extends keyof T[P]
    ? ReturnType<NonNullable<T[P]['validation']>> extends RuleForRequired
      ? RequiredField<FieldType<T[P]>>
      : OptionalField<FieldType<T[P]>>
    : OptionalField<FieldType<T[P]>>
}

// eslint-disable-next-line
export function extractType<T extends SanityDocumentCreator>(_schema: T) {
  return {} as SanityDocumentObject<T>
}

export function createSanityDocument<T extends SanityDocumentCreator>(
  documentTitle: string,
  schema: T
): DocumentField {
  return {
    name: documentTitle,
    title: splitStringByCase(documentTitle),
    type: 'document' as const,
    fields: Object.entries(schema).map(([key, value]) => ({
      // eslint-disable-next-line
      // @ts-ignore
      title: splitStringByCase(key),
      ...value,
      name: key,
    })),
  }
}

interface ExtraFields<T extends string = string> {
  _createdAt: {
    type: 'string'
  }
  _updatedAt: {
    type: 'string'
  }
  _id: {
    type: 'string'
  }
  _rev: {
    type: 'string'
  }
  _type: {
    type: '_type'
    name: T
  }
}

export function extraFields<T extends string>(name: T) {
  return {
    _createdAt: {
      type: 'string',
    },
    _updatedAt: {
      type: 'string',
    },
    _id: {
      type: 'string',
    },
    _rev: {
      type: 'string',
    },
    _type: {
      type: '_type',
      name,
    },
  }
}

export function createSchema<
  Schema extends SanityDocumentCreator,
  SchemaName extends string
>(documentTitle: SchemaName, schema: Schema) {
  return {
    schema,
    builder: new QueryBuilder(
      schema as Schema & ExtraFields<SchemaName>,
      documentTitle
    ),
    type: extractType(schema as Schema & ExtraFields<SchemaName>),
    document: createSanityDocument(documentTitle, schema),
  }
}
