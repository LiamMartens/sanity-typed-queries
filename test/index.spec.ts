import { defineDocument } from '../src'
import { capitalise, splitStringByCase } from '../src/utils'
import { defineFields } from '../src/extractor'

const { document, builder } = defineDocument('typeOfDocument', {
  name: {
    type: 'string',
    validation: Rule => Rule.required(),
  },
  postcode: {
    type: 'string',
    validation: Rule => Rule.required(),
  },
  cost: {
    type: 'number',
  },
  description: {
    type: 'text',
    rows: 2,
    validation: Rule => Rule.required(),
  },
})

describe('schema creator', () => {
  test('generates correct schema fields for document', () => {
    expect(document).toMatchSnapshot()
  })

  test('generates correct schema fields when fields are defined', () => {
    expect(defineFields({ field1: { type: 'string' } })).toMatchSnapshot()
  })

  test('generates correct queries', () => {
    expect(builder.first().use()[0]).toMatchSnapshot()
    expect(builder.pick('description').use()[0]).toMatchSnapshot()
    expect(builder.pick(['description']).use()[0]).toMatchSnapshot()
    expect(builder.select(0, 2).use()[0]).toMatchSnapshot()
    expect(builder.select(0, 2, true).use()[0]).toMatchSnapshot()
    expect(builder.orderBy('postcode', 'desc').use()[0]).toMatchSnapshot()
    expect(builder.pick(['_type']).orderBy('_type').use()[0]).toMatchSnapshot()
  })

  test('generates custom mappings correctly', () => {
    expect(
      builder
        .map(r => ({ mappedItem: r._type.use() }))
        .pick(['mappedItem'])
        .use()[0]
    ).toMatchSnapshot()
    expect(
      builder
        .map(r => ({ mappedItem: r._type.use() }))
        .pick('mappedItem')
        .use()[0]
    ).toMatchSnapshot()
    expect(
      builder
        .map(r => ({ mappedItem: r._type.use() }))
        .pick('mappedItem')
        .first()
        .use()[0]
    ).toMatchSnapshot()
    expect(
      builder
        .map({ mappedItem: 'customMap->resolve.thing' })
        .pick(['mappedItem'])
        .use()[0]
    ).toMatchSnapshot()
  })
})

describe('utils', () => {
  test('splits schema names correctly', () => {
    expect(splitStringByCase('testCMS')).toEqual('Test CMS')
    expect(splitStringByCase('testCMSToday')).toEqual('Test CMS Today')
  })

  test('capitalises names correctly', () => {
    expect(capitalise('johnson')).toEqual('Johnson')
  })
})
