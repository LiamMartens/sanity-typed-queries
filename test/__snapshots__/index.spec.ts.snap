// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema creator generates correct queries 1`] = `"*[_type == 'typeOfDocument'][0]"`;

exports[`schema creator generates correct queries 2`] = `"*[_type == 'typeOfDocument'].description"`;

exports[`schema creator generates correct queries 3`] = `"*[_type == 'typeOfDocument'] { description }"`;

exports[`schema creator generates correct queries 4`] = `"*[_type == 'typeOfDocument'][0..2]"`;

exports[`schema creator generates correct queries 5`] = `"*[_type == 'typeOfDocument'][0...2]"`;

exports[`schema creator generates correct queries 6`] = `"*[_type == 'typeOfDocument'] | order(postcode desc)"`;

exports[`schema creator generates correct queries 7`] = `"*[_type == 'typeOfDocument'] { _type } | order(_type asc)"`;

exports[`schema creator generates correct schema fields for document 1`] = `
Object {
  "fields": Array [
    Object {
      "name": "name",
      "title": "Name",
      "type": "string",
      "validation": [Function],
    },
    Object {
      "name": "postcode",
      "title": "Postcode",
      "type": "string",
      "validation": [Function],
    },
    Object {
      "name": "cost",
      "title": "Cost",
      "type": "number",
    },
    Object {
      "name": "description",
      "rows": 2,
      "title": "Description",
      "type": "text",
      "validation": [Function],
    },
  ],
  "name": "typeOfDocument",
  "title": "Type Of Document",
  "type": "document",
}
`;

exports[`schema creator generates correct schema fields when fields are defined 1`] = `
Array [
  Object {
    "name": "field1",
    "title": "Field1",
    "type": "string",
  },
]
`;

exports[`schema creator generates custom mappings correctly 1`] = `"*[_type == 'typeOfDocument'] { \\"mappedItem\\": _type }"`;

exports[`schema creator generates custom mappings correctly 2`] = `"*[_type == 'typeOfDocument']._type"`;

exports[`schema creator generates custom mappings correctly 3`] = `"*[_type == 'typeOfDocument'][0]._type"`;

exports[`schema creator generates custom mappings correctly 4`] = `"*[_type == 'typeOfDocument'] { \\"mappedItem\\": customMap->resolve.thing }"`;

exports[`schema creator generates custom mappings correctly 5`] = `"*[_type == 'typeOfDocument'] { ..., \\"mappedItem\\": customMap->resolve.thing }"`;
